### PR TITLE
feat(view): Floating nvim tree window #1377

### PR DESCRIPTION
**Pull request by Khrees**
This PR adds option to display nvim-tree window as float. Requires setting `nvim-tree.actions.open_file.quit_on_open` to `true` to work properly.

demo

References #1377

### CODE REVIEW

**Review by kyazdani42 — CHANGES_REQUESTED** [review]

**Comment by kyazdani42** [comment]
thanks for the PR :) i think this was asked a lot by people, we could ask them to try it out.
https://github.com/kyazdani42/nvim-tree.lua/issues/135

**Comment by alex-courtis** [comment]
I'll give this a lot of testing before review/approve.

There are a lot of cases like `update_focused_file` that need testing.

**Comment by alex-courtis** [comment]
Switching away from the nvim-tree window is problematic as the cursor shows in the nvim-tree window. This is a confusing user experience.
<details><summary>Minimal config with quit_on_open and float.enable set</summary><br>


</details>

`<C-W>w` was used to switch windows.

**Practical Proposal**
quit_on_open is insufficient as it only covers one specific case.
Close the nvim-tree window when focus is lost.
<details><summary>gitsigns.nvim behaviour</summary><br>

1377-gitsigns
</details>

File: doc/nvim-tree-lua.txt
Comment (alex-courtis) [reply doc/nvim-tree-lua.txt:200]: If we require this, `quit_on_open` should be overridden and set when `float.enable` is set.

File: lua/nvim-tree.lua
Comment (alex-courtis) [inline lua/nvim-tree.lua:454]: `anchor` would be useful.

File: doc/nvim-tree-lua.txt
Comment (alex-courtis) [reply doc/nvim-tree-lua.txt:200]: If we require this, `quit_on_open` should be overridden and set when `float.enable` is set.

**Review by alex-courtis — CHANGES_REQUESTED** [review]
Very promising, needs UX improvements: https://github.com/kyazdani42/nvim-tree.lua/pull/1462#issuecomment-1194757935

I'm tempted to release this early and resolve bugs / improve UX after users raise issues; I expect there will be a lot.

@Khrees can you commit to ongoing support / fixes / changes for this feature?

**Comment by kyazdani42** [comment]
I agree we could merge a first version and improve on that. I'm not sure all the UX impacts that will raise.
Closing the window when leaving it would also be a good addition i think

**Comment by Khrees** [comment]
> @Khrees can you commit to ongoing support / fixes / changes for this feature?

Yes, but it will be somewhat limited.

File: lua/nvim-tree.lua
Comment (Khrees) [reply lua/nvim-tree.lua:454]: Done.

File: doc/nvim-tree-lua.txt
Comment (Khrees) [reply doc/nvim-tree-lua.txt]: `quit_on_open` is now enforced if `float.enable` is set.

File: doc/nvim-tree-lua.txt
Comment (kyazdani42) [inline doc/nvim-tree-lua.txt:203]: please document these option. See `:help nvim-tree.view` force instance to see how its done :)

**Review by kyazdani42 — CHANGES_REQUESTED** [review]

File: doc/nvim-tree-lua.txt
Comment (Khrees) [reply doc/nvim-tree-lua.txt]: Done.

**Review by kyazdani42 — APPROVED** [review]
LGTM, lets wait for alex's review

**Review by alex-courtis — CHANGES_REQUESTED** [review]
- [ ] nvim-tree float is still not closing on window exit

<details><summary>minimal config</summary><br>


</details>

- [ ] open up `nvim_open_win` config

After playing with this for a while, it became clear that the user needs to be able to set all of the config passed to `nvim_open_win`, rather than the limited config we allow.

Solution is to allow the user to pass all config, rather than just what we specify in the defaults. The user may pass an invalid config to `nvim_open_win` however they will get a useful error message.

Added `float.open_win_config` which is no longer validated, however it is merged. Simplified help to match.

Tidied view.lua: just pass the options to open_win.

[0001-add-view.float.open_win_config-skipping-validation.patch.txt](https://github.com/kyazdani42/nvim-tree.lua/files/9227988/0001-add-view.float.open_win_config-skipping-validation.patch.txt)

**Comment by Khrees** [comment]
@alex-courtis I have applied your patch and made nvim-tree window close on focus lost.

**Comment by alex-courtis** [comment]
> @alex-courtis I have applied your patch and made nvim-tree window close on focus lost.

Thank you. I can't test / approve until the weekend.

**Review by alex-courtis — APPROVED** [review]

**Comment by alex-courtis** [comment]
Works beautifully with a variety of options.

Let's see how users like it...

**Comment by davidsierradz** [comment]
Hi! Testing and it works super well. Could we allow width, height, col and row to be functions that return numbers? So we can calculate on demand this values. My use-case would be to open the float window in the center, with something like:
```lua
local screen_w = vim.opt.columns:get()
local screen_h = vim.opt.lines:get() - vim.opt.cmdheight:get()
local _width = screen_w
local _height = screen_h
local width = math.floor(_width)
local height = math.floor(_height)
local center_y = (vim.opt.lines:get() - _height) / 2
local center_x = (screen_w - _width) / 2
```
Thanks!

**Comment by alex-courtis** [comment]
> Hi! Testing and it works super well. Could we allow width, height, col and row to be functions that return numbers? So we can calculate on demand this values. My use-case would be to open the float window in the center, with something like:
> 
> ```lua
> local screen_w = vim.opt.columns:get()
> local screen_h = vim.opt.lines:get() - vim.opt.cmdheight:get()
> local _width = screen_w
> local _height = screen_h
> local width = math.floor(_width)
> local height = math.floor(_height)
> local center_y = (vim.opt.lines:get() - _height) / 2
> local center_x = (screen_w - _width) / 2
> ```
> 
> Thanks!

Very interesting. Rather than individual functions, we could instead have `view.float.open_win_config` be the function. Please raise a [feature request](https://github.com/kyazdani42/nvim-tree.lua/issues/new?assignees=&labels=feature+request&template=feature_request.md&title=).

**Comment by avalonv** [comment]
Is there a summary for how to enable and test this for someone who just installed the extension?

**Comment by alex-courtis** [comment]
> Is there a summary for how to enable and test this for someone who just installed the extension?

`:help nvim-tree.view.float`

**Pull request merged**
